shell: kernel: group thread-related cmds & add CPU mask affinity/pinning cmds

ycsin opened the pull request:
Group `kernel threads` & `kernel stacks` under `thread` subcmd as `kernel thread list` & `kernel thread stacks`, then add CPU mask affinity/pinning commands.

ycsin committed:
shell: modules: kernel: group thread-related cmds

`stacks` `threads` & `unwind` are commands related to kernel
thread, group them one level inside a main `thread` command,
and rename `threads` to `list`:

- `kernel threads` -> `kernel thread list`
- `kernel stacks` -> `kernel thread stacks`
- `kernel unwind` -> `kernel thread unwind`

Additionally, rename and mark the `thread_valid()` function
as `__maybe_unused` as it might be unused for architectures
without stackwalk implementation, and use the locked version
of the thread iterator (`k_thread_foreach()`)

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
shell: modules: kernel: add additional check in unwind cmd

Use `shell_strtoull()` to parse the thread ID argument for
error checking.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
shell: modules: kernel: add CPU mask affinity/pinning commands

Added commands to access the APIs in the kernel/cpu_mask.c

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

jakub-uC commented on subsys/shell/modules/kernel_service.c:
```diff
@@ -333,8 +271,269 @@ static int cmd_kernel_stacks(const struct shell *sh,
+	SHELL_CMD_ARG(unwind, NULL, "Unwind a thread.", cmd_kernel_thread_unwind, 1, 1),
+#endif /* CONFIG_ARCH_HAS_STACKWALK */
+#ifdef CONFIG_SCHED_CPU_MASK
+	SHELL_CMD_ARG(mask, &sub_kernel_thread_mask, "Configure thread CPU mask affinity.",
```
Did you consider using`SHELL_COND_CMD_ARG_REGISTER` instead of `ifdef`?

ycsin replied:
No, ain't aware of that, I pushed another version with `SHELL_COND_CMD_ARG`, PTAL

ycsin replied:
Reverted back from `SHELL_COND_CMD_ARG` to `SHELL_CMD_ARG` because the former requires the command to be there and thus the `#ifdef` can't be avoided, they just relocated

kartben commented:
Not opposed to the change but please document the change in the release notes

ycsin committed:
doc: updated release note & migration guide for the `kernel thread` cmd

Added entries in the release note & migration guide about the
changes to the `kernel thread` shell command.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented:
> Not opposed to the change but please document it in the release notes

Done!

ycsin commented:
force pushed to fix merge conflict

ycsin committed:
shell: modules: kernel: fix typo in `cmd_kernel_uptime()`

Caught by CI in another PR, should be 'Unsupported' instead.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>
